### PR TITLE
[PLAYER-5330] Add handling and passing in core information about ui visibility.

### PR DESCRIPTION
**Pull request by atopilskii**
**Task:** It was neccessary to add callback in Android SDK when scrubber bar visibility is changed 
**Actions:** I move logic related to calculating if ui is visible from render() to componentDidUpdate() and in case if visibility is not equals previous state - pass this information to core. For cast and audio only views I add passing 'true' after mount view. 
**Result** We can listen this event in OoyalaSkinLayoutController, sample is here: https://github.com/ooyala/android-sample-apps/compare/candidate...PLAYER-5330

### CODE REVIEW

File: sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js [hunk 1]
```diff
@@ -244,13 +246,22 @@ export default class AdPlaybackScreen extends Component {
     return iconViews;
   };
 
-  render() {
+  componentDidUpdate(prevProps: Readonly<P>, prevState: Readonly<S>, snapshot: SS) {
```
Comment (ooyala-amohnatkin) [inline sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js]: where do these types come from?

File: sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js [hunk 2]
```diff
@@ -244,13 +246,22 @@ export default class AdPlaybackScreen extends Component {
...
     const isPastAutoHideTime = (new Date).getTime() - this.props.lastPressedTime > AUTOHIDE_DELAY;
     const doesAdRequireControls = this.props.ad && this.props.ad.requireControls;
     // TODO: IMA Ads UI is still not supported - No way to show UI while allowing Learn More in a clean way
     const isContent = !this.props.ad;
-    const shouldShowControls = this.props.screenReaderEnabled ? true : !isPastAutoHideTime && (doesAdRequireControls || isContent);
+    const isVisible = this.props.screenReaderEnabled ? true : !isPastAutoHideTime && (doesAdRequireControls || isContent);
 
+    if (isVisible !== this.state.shouldShowControls) {
```
Comment (ooyala-amohnatkin) [inline sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js]: not clear why do you use state here

File: sdk/android/skin/src/main/java/com/ooyala/android/skin/OoyalaSkinLayoutController.java [hunk 3]
```diff
@@ -774,4 +777,22 @@ private Activity getActivity() {
...
+   * This listener will be cleaned in {@link #destroy} method
+   * @param onVisibilityControlsChangeListener Consumer that handle changing of visibility skin ui
+   */
+  public void addOnVisibilityControlsChangeListener(Consumer<Boolean> onVisibilityControlsChangeListener) {
+    this.onVisibilityControlsChangeListener = onVisibilityControlsChangeListener;
+  }
+
+  /**
+   * Handles changing of skin controls visibility from {@link OoyalaReactBridge}.
+   * @param isVisible is true if controls visible, is false if controls not visible
```
Comment (AlinaVoronkovaEpam) [inline sdk/android/skin/src/main/java/com/ooyala/android/skin/OoyalaSkinLayoutController.java]: `if controls are/aren't visible`

File: sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js (hunk 1 above)
Comment (atopilskii) [reply sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js]: It auto-generated code, if you can suggest me how write this method correctly i would be glad.

File: sdk/android/skin/src/main/java/com/ooyala/android/skin/OoyalaSkinLayoutController.java [hunk 4]
```diff
@@ -774,4 +777,22 @@ private Activity getActivity() {
     }
   }
 
+  /**
+   * Provide possibility to handle changing of skin controls visibility.
+   * This listener will be cleaned in {@link #destroy} method
```
Comment (AlinaVoronkovaEpam) [inline sdk/android/skin/src/main/java/com/ooyala/android/skin/OoyalaSkinLayoutController.java]: `This listener will be released` sounds much more naturally for me. I don't insist

File: sdk/android/skin/src/main/java/com/ooyala/android/skin/OoyalaSkinLayoutController.java [hunk 5]
```diff
@@ -774,4 +777,22 @@ private Activity getActivity() {
...
+    this.onVisibilityControlsChangeListener = onVisibilityControlsChangeListener;
+  }
+
+  /**
+   * Handles changing of skin controls visibility from {@link OoyalaReactBridge}.
+   * @param isVisible is true if controls visible, is false if controls not visible
+   */
+  void onVisibilityControlsChanged(boolean isVisible) {
+    if (onVisibilityControlsChangeListener != null) {
+      onVisibilityControlsChangeListener.accept(isVisible);
```
Comment (AlinaVoronkovaEpam) [inline sdk/android/skin/src/main/java/com/ooyala/android/skin/OoyalaSkinLayoutController.java:805]: :)

File: sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js (hunk 2 above)
Comment (atopilskii) [reply sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js]: i check previous state to learn if i have to pass information about ui

File: sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js (hunk 1 above)
Comment (ooyala-amohnatkin) [reply sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js]: this file is not covered with Flow, no types needed

Comment (atopilskii) [reply sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js]: deleted, thanks!

File: sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js [hunk 6]
```diff
@@ -265,8 +276,8 @@ export default class AdPlaybackScreen extends Component {
         <View style={styles.adContainer}>
           {adBar}
           {this._renderPlaceholder(adIcons)}
-          {this._renderPlayPause(shouldShowControls)}
-          {this._renderBottomOverlay(shouldShowControls)}
+          {this._renderPlayPause(this.state.shouldShowControls)}
```
Comment (ooyala-amohnatkin) [inline sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js]: `_renderPlayPause` and `_renderBottomOverlay` have access to the state, no need to pass it as an argument

File: sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js [hunk 7]
```diff
@@ -56,7 +57,8 @@ export default class AdPlaybackScreen extends Component {
   };
 
   state = {
-    showControls: true
+    showControls: true,
```
Comment (ooyala-amohnatkin) [inline sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js]: `showControls` is not modified anywhere and is always `true`, so it's redundant

**Comment by SergeyBicarte** [comment]
It's already seems good solution, but it could be slightly improved, I suggest to add addOnVisibilityControlsChangeListener in a LayoutController interface, so we or user can implement it in any type of layout (native, RN or custom)

File: sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js [hunk 8]
```diff
@@ -40,6 +40,7 @@ export default class AdPlaybackScreen extends Component {
       onScrub: PropTypes.func,
       handleVideoTouch: PropTypes.func,
       handleControlsTouch: PropTypes.func,
+      onControlsVisibilityChanged: PropTypes.func,
```
Comment (loginov-rocks) [inline sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js]: If this passed every time and cannot be undefined, then we have to make this `.isRequired`

File: sdk/react/src/views/CastConnectedScreen/CastConnectedScreen.js [hunk 9]
```diff
@@ -56,6 +57,14 @@ export default class CastConnectedScreen extends Component {
     markers: PropTypes.array.isRequired,
   };
 
+  state = {
+    shouldShowControls: true
```
Comment (loginov-rocks) [inline sdk/react/src/views/CastConnectedScreen/CastConnectedScreen.js]: It looks like this state never changes, so we can get rid of it. It's a good idea to make components stateless everywhere it's possible.

File: sdk/react/src/views/VideoView/VideoView.js [hunk 10]
```diff
@@ -408,9 +424,9 @@ export default class VideoView extends Component {
         {this._renderPlaceholder()}
         {this._renderBottom()}
         {this._renderAdOverlay()}
-        {this._renderPlayPause(shouldShowControls)}
+        {this._renderPlayPause(this.state.shouldShowControls)}
```
Comment (loginov-rocks) [inline sdk/react/src/views/VideoView/VideoView.js]: The same not as @ooyala-amohnatkin made: we can get this state value from the method itself.

**Review by loginov-rocks — CHANGES_REQUESTED** [review]
Please review a few comments about JS.

File: sdk/react/src/Core.js [hunk 11]
```diff
@@ -211,6 +211,10 @@ export default class Core {
     }
   }
 
+  onControlsVisibilityChanged(isVisible) {
+    this.bridge.onVisibilityControlsChanged({ isVisible: isVisible });
```
Comment (loginov-rocks) [inline sdk/react/src/Core.js]: You can use shorthand here: `{ isVisible }` - strange that ESLint didn't mention that.

**Review by ooyala-amohnatkin — APPROVED** [review]

File: sdk/react/src/Core.js (hunk 11 above)
Comment (atopilskii) [reply sdk/react/src/Core.js]: done

File: sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js (hunk 6 above)
Comment (atopilskii) [reply sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js]: done

File: sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js (hunk 7 above)
Comment (atopilskii) [reply sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js]: done

File: sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js (hunk 8 above)
Comment (atopilskii) [reply sdk/react/src/views/AdPlaybackScreen/AdPlaybackScreen.js]: done, thanks!

File: sdk/react/src/views/CastConnectedScreen/CastConnectedScreen.js (hunk 9 above)
Comment (atopilskii) [reply sdk/react/src/views/CastConnectedScreen/CastConnectedScreen.js]: done, thanks!

File: sdk/react/src/views/VideoView/VideoView.js (hunk 10 above)
Comment (atopilskii) [reply sdk/react/src/views/VideoView/VideoView.js]: done

**Review by loginov-rocks — APPROVED** [review]
LGTM from JS side. Let's wait iOS for approval.

File: sdk/android/skin/src/main/java/com/ooyala/android/skin/OoyalaSkinLayoutController.java (hunk 5 above)
Comment (atopilskii) [reply sdk/android/skin/src/main/java/com/ooyala/android/skin/OoyalaSkinLayoutController.java:805]: ? :)

File: sdk/android/skin/src/main/java/com/ooyala/android/skin/OoyalaSkinLayoutController.java (hunk 3 above)
Comment (atopilskii) [reply sdk/android/skin/src/main/java/com/ooyala/android/skin/OoyalaSkinLayoutController.java]: done, thanks!

**Review by SergeyBicarte — APPROVED** [review]

File: sdk/android/skin/src/main/java/com/ooyala/android/skin/OoyalaSkinLayoutController.java (hunk 4 above)
Comment (atopilskii) [reply sdk/android/skin/src/main/java/com/ooyala/android/skin/OoyalaSkinLayoutController.java]: not fixed since it was removed

**Review by maxkupetskii — APPROVED** [review]

**Pull request merged**